src/Logary/Internals/Service.fs

```diff
@@ -0,0 +1,115 @@
+namespace Logary.Internals
+
+open NodaTime
+open Hopac
+open Hopac.Infixes
+open Logary
+open Logary.Internals
+open Logary.Message
+
+type ServiceState =
+  | Initialising
+  | Running
+  | Paused
+  | ShuttingDown
+  | Shutdown
+  | Faulted of unhandled:exn // the supervised job faulted
+
+type Service =
+  abstract name : string
+
+/// A service is an abstract interface that encapsulates a service
+/// implementation.
+type Service<'i> =
+  inherit Service
+  abstract instance : 'i
+
+and ShutdownService =
+  /// Gets the list of service states and the duration that the service was in
+  /// each of them.
+  abstract transitions : (ServiceState * Duration) list
+
+module Service =
+
+  type T =
+    private {
+      name : PointName
+      pauseCh : Ch<Ch<unit> * Promise<unit>>
+      resumeCh : Ch<Ch<unit> * Promise<unit>>
+      shutdown : Alt<Promise<unit>>
+      getStateCh : Ch<IVar<ServiceState>>
+      ilogger : Logger
+    }
+  with
+    override x.ToString() =
+      sprintf "Service(%O)" x.name
+
+  let pause (x : Service<T>) : Alt<unit> =
+    x.instance.pauseCh *<+->- fun ack nack -> ack, nack
+
+  let resume (x : Service<T>) : Alt<unit> =
+    x.instance.resumeCh *<+->- fun ack nack -> ack, nack
+
+  let shutdown (x : Service<T>) : Alt<Promise<unit>> =
+    x.instance.shutdown
+
+  let getState (t : Service<T>) : Alt<ServiceState> =
+    t.instance.getStateCh *<+=>- id
+
+  /// Create a new service from a supervised job. It's up to the caller to
+  /// decide what sort of policy the supervised job should have.
+
+  let create (ilogger : Logger)
+             name
+             (pause : Alt<_>)
+             (resume : Alt<_>)
+             (shutdown : Alt<_>)
+             (server : SupervisedJob<unit>)
+             : Job<Service<T>> =
+    let ilogger = ilogger |> Logger.apply (setSimpleName (sprintf "Logary.Service(%s)" name))
+    let pn = PointName.ofSingle name
+    let mserver = memo server
+    let pauseCh, resumeCh, getStateCh = Ch (), Ch (), Ch ()
+
+    let rec running transitions =
+      Alt.choose [
+        pauseCh ^=>. pause
```

Comment:
```fsharp
module internal GlobalService =
  open Global

  let create (t : T) (ilogger : Logger) =
    let logger = ilogger |> Logger.apply (setSimpleName "Logary.Globals")
    let pauseCh, resumeCh, shutdownCh = Ch (), Ch (), Ch ()

    let rec init () =
      let prev = !config
      initialise t
      running t (fst prev)

    and running myself prev =
      Alt.choose [
        pauseCh ^=> fun (ack, nack) ->
          logger.debug (eventX "Pausing.")
          initialise prev
          ack *<= () >>=. running myself prev

        resumeCh ^=> fun (ack, nack) ->
          logger.debug (eventX "Resuming.")
          initialise myself
          ack *<= () >>=. running myself prev

        shutdownCh ^=> fun ack ->
          logger.debug (eventX "Shutting down.")
          initialise prev
          ack *<= ()
      ]

    let shutdown : Alt<Promise<unit>> =
      shutdownCh *<-=>= fun repl -> repl
      |> Alt.afterFun (fun iv -> iv :> _)

    let loop = Job.supervise logger Policy.terminate (init ())
    Service.create logger "globals" pauseCh resumeCh shutdown loop
```

if the `pause` is the function , the last line above passing an `channel` to it, is this work in progress，just for temporary compile ?